clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/id/DcosSpinnakerAppId.groovy

```diff
@@ -0,0 +1,153 @@
+package com.netflix.spinnaker.clouddriver.dcos.deploy.util.id
+
+import static com.google.common.base.Preconditions.checkArgument
+import static com.google.common.base.Strings.nullToEmpty
+
+import com.netflix.frigga.Names
+
+/**
+ * Represents a hierarchical Spinnaker specific application identifier for DCOS.
+ * Structure - /account/region/app-stack-detail-sequence
+ */
+class DcosSpinnakerAppId {
+    public final static def SAFE_REGION_SEPARATOR = "_"
+
+    private final def marathonPath
+
+    public DcosSpinnakerAppId(final String marathonAppId) {
```

Comment:
Generally I prefer static factory methods over constructors when you have multiple ways of creating an object. It's the first suggestion in Effective Java, and he lists the pros better than I can (http://www.informit.com/articles/article.aspx?p=1216151). 

At the very least, I would argue that we shouldn't have both static factory methods and public constructors as it stands right now. It should be consistent one way or the other (my vote is for static factory methods). 

Do we actually need the Optional form of construction? I don't recall that was necessary before - anywhere we were constructing a DcosSpinnakerId before (from a marathon id), we were validating it beforehand to make sure we weren't passing an "illegal" id to the static factory method.